### PR TITLE
[WIP] Implement 2nd pass training using 1-best decoding results from the 1st pass network

### PR DESCRIPTION
**Pull request by csukuangfj**
It implements https://github.com/k2-fsa/snowfall/pull/106#issuecomment-803796177

> BTW, since it seems this is hard to get to work, if you feel like it you could work on a simpler idea.
In training time we'd take the best-path alignment and using some kind of RNN or masked attention
we'd predict the next label in that best-path. (We'd probably take in the output of the 1st network
as an input to that). The sequence length here is the same as the original sequence length.
In test time the way this would work at least initially, is we'd run this on n-best lists obtained from the
1st-pass decoding and use the scores to decide which of the n-best paths to keep. There are more accurate
decoding methods we could look into later.

---

The training objf is decreasing and seems to be converging. Will post the decoding results later.

### CODE REVIEW

File: snowfall/models/second_pass_model.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+                          mask=attn_mask,
+                          src_key_padding_mask=key_padding_mask)
+
+        # x2 is (T, B, F)
+
+        x2 = x2.permute(1, 0, 2)
+
+        # now x2 is (B, T, F)
+
+        x_concat = torch.cat((padded_acoustics, x2), dim=-1)
```
Comment (csukuangfj) [inline snowfall/models/second_pass_model.py:217]: TODO(fangjun): Use cross attention here, like decoders
- query: x2
- key and value: padded_acoustics

File: snowfall/common2.py [hunk 2]
```diff
@@ -0,0 +1,484 @@
+#!/usr/bin/env python3
```
Comment (csukuangfj) [inline snowfall/common2.py:1]: common2.py is the same as common.py, except that it has some code supporting
the second pass model. To avoid conflicts with the master,  a new file is used.

The same goes for the following xxx2.py files, e.g., lm_rescore2.py, mmi2.py.

File: snowfall/decoding/second_pass.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+# This file implements the ideas proposed by Daniel Povey.
+#
+# See https://github.com/k2-fsa/snowfall/issues/232 for more details
+#
+
+import k2
+
+
+class Nbest(object):
```
Comment (csukuangfj) [inline snowfall/decoding/second_pass.py:11]: @danpovey 

This file implements the `Nbest` class proposed in
https://github.com/k2-fsa/snowfall/issues/232#issue-941386780

Comment (danpovey) [reply snowfall/decoding/second_pass.py:11]: That's great!  Yes it looks like what I had in mind.
I assume you would separate it from this PR though?  Or maybe even submit it to k2?  Since there's a lot going on here.

Comment (csukuangfj) [reply snowfall/decoding/second_pass.py:11]: Will move it to k2.
